### PR TITLE
enable prebuilt model

### PR DESCRIPTION
**Pull request by ylwu-amzn**
### Description
As we have published prebuilt models to OpeSearch repo, we can enable uploading prebuilt model function now.

Example:
```
POST /_plugins/_ml/models/_upload
{
  "name": "huggingface/sentence-transformers/paraphrase-MiniLM-L3-v2",
  "version": "1.0.1",
  "model_format": "TORCH_SCRIPT"
}
```
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: common/src/main/java/org/opensearch/ml/common/transport/upload/MLUploadInput.java [hunk 1]
```diff
@@ -74,23 +74,17 @@ public MLUploadInput(FunctionName functionName,
             throw new IllegalArgumentException("model version is null");
         }
         //TODO: enable prebuilt model in 2.6
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/upload/MLUploadInput.java:76]: May be we can remove this line?

File: common/src/main/java/org/opensearch/ml/common/model/TextEmbeddingModelConfig.java [hunk 2]
```diff
@@ -184,7 +184,7 @@ public String getName() {
 
         public static PoolingMode from(String value) {
             try {
-                return PoolingMode.valueOf(value);
+                return PoolingMode.valueOf(value.toUpperCase(Locale.ROOT));
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/model/TextEmbeddingModelConfig.java:187]: Not this line.

 do we need to change `weightedmean` value?

File: common/src/main/java/org/opensearch/ml/common/transport/upload/MLUploadInput.java [hunk 3]
```diff
@@ -74,23 +74,17 @@ public MLUploadInput(FunctionName functionName,
...
-//        if (url != null) {
-//            if (modelFormat == null) {
-//                throw new IllegalArgumentException("model format is null");
-//            }
-//            if (modelConfig == null) {
-//                throw new IllegalArgumentException("model config is null");
-//            }
-//        }
+        if (url != null) {
+            if (modelFormat == null) {
```
Comment (dhrubo-os) [inline common/src/main/java/org/opensearch/ml/common/transport/upload/MLUploadInput.java:78]: Why do we need to check `modelFormat` two times?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/MLEngine.java [hunk 4]
```diff
@@ -34,13 +37,18 @@ public MLEngine(Path opensearchDataFolder) {
         djlModelsCachePath = djlCachePath.resolve("models_cache");
     }
 
-    public String getCIPrebuiltModelConfigPath(String modelName, String version) {
-        return String.format("https://ci.opensearch.org/ci/dbc/models/ml-models/%s/%s/config.json", modelName, version, Locale.ROOT);
+    public String getPrebuiltModelConfigPath(String modelName, String version, MLModelFormat modelFormat) {
+        String format = modelFormat.name().toLowerCase(Locale.ROOT);
+        return String.format("%s/%s/%s/%s/config.json", MODEL_REPO, modelName, version, format, Locale.ROOT);
```
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/MLEngine.java:42]: Should we check if the file is actually there? May be raise an exception if the model config path is not valid?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/MLEngine.java [hunk 5]
```diff
@@ -34,13 +37,18 @@ public MLEngine(Path opensearchDataFolder) {
...
-    public String getCIPrebuiltModelPath(String modelName, String version) {
-        int index = modelName.lastIndexOf("/") + 1;
-        return String.format("https://ci.opensearch.org/ci/dbc/models/ml-models/%s/%s/%s.zip", modelName, version, modelName.substring(index), Locale.ROOT);
+    public String getPrebuiltModelPath(String modelName, String version, MLModelFormat modelFormat) {
+        int index = modelName.indexOf("/") + 1;
+        // /huggingface/sentence-transformers/msmarco-distilbert-base-tas-b/1.0.0/onnx/sentence-transformers_msmarco-distilbert-base-tas-b-1.0.0-torch_script.zip
+        // /huggingface/sentence-transformers/msmarco-distilbert-base-tas-b/1.0.0/onnx/config.json
+        String format = modelFormat.name().toLowerCase(Locale.ROOT);
+        String modelZipFileName = modelName.substring(index).replace("/", "_") + "-" + version + "-" + format;
+        return String.format("%s/%s/%s/%s/%s.zip", MODEL_REPO, modelName, version, format, modelZipFileName, Locale.ROOT);
```
Comment (dhrubo-os) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/MLEngine.java:51]: Same here, may be we can check if the file is actually there? if not raise an exception saying invalid file path?

**Review by dhrubo-os — COMMENTED** [review]
Don't we need to add any test for this?

**Review by Zhangxunmt — APPROVED** [review]

File: common/src/main/java/org/opensearch/ml/common/model/TextEmbeddingModelConfig.java (hunk 2 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/model/TextEmbeddingModelConfig.java:187]: I prefer to change it in a separate PR. That change takes some time to test.

File: common/src/main/java/org/opensearch/ml/common/transport/upload/MLUploadInput.java (hunk 3 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/transport/upload/MLUploadInput.java:78]: Good catch, I will remove this.

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/MLEngine.java (hunk 4 above)
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/MLEngine.java:42]: The download function will throw exception if file doesn't exist.

File: common/src/main/java/org/opensearch/ml/common/transport/upload/MLUploadInput.java (hunk 1 above)
Comment (ylwu-amzn) [reply common/src/main/java/org/opensearch/ml/common/transport/upload/MLUploadInput.java:76]: yes, will remove this

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/MLEngine.java (hunk 5 above)
Comment (ylwu-amzn) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/MLEngine.java:51]: The download function will throw exception if file doesn't exist.

**Comment by dhrubo-os** [comment]
I had another comment about testing. Don't we need any corresponding testing for this code?

**Comment by ylwu-amzn** [comment]
> I had another comment about testing. Don't we need any corresponding testing for this code?

This PR is to enable pre-built model feature. 
The code was developed in this PR #655 , as we have no model published on model repo, we disabled in PR #682 before 2.5 release.
You can find testing code in PR #655. As the Github CI instance is not so powerful, we don't add IT for prebuilt model now. We have called out this issue to infra on 2.5 retro. Create an [issue](https://github.com/opensearch-project/ml-commons/issues/732) to track this

**Review by dhrubo-os — APPROVED** [review]

**Review by Zhangxunmt — APPROVED** [review]

**Review by dhrubo-os — APPROVED** [review]

**Pull request merged**